src/Common/ThreadPool.h

```diff
@@ -162,20 +162,19 @@ class ThreadFromGlobalPool : boost::noncopyable
 
     template <typename Function, typename... Args>
     explicit ThreadFromGlobalPool(Function && func, Args &&... args)
-        : state(std::make_shared<Poco::Event>())
-        , thread_id(std::make_shared<std::thread::id>())
+        : state(std::make_shared<State>())
     {
-        /// NOTE: If this will throw an exception, the destructor won't be called.
+        /// NOTE:
+        /// - If this will throw an exception, the destructor won't be called
+        /// - this pointer cannot be passed in the lambda, since after detach() it will not be valid
         GlobalThreadPool::instance().scheduleOrThrow([
-            thread_id = thread_id,
             state = state,
             func = std::forward<Function>(func),
             args = std::make_tuple(std::forward<Args>(args)...)]() mutable /// mutable is needed to destroy capture
         {
-            auto event = std::move(state);
-            SCOPE_EXIT(event->set());
+            SCOPE_EXIT(state->event.set());
```

Comment:
I think this line was to prevent storing some capture after the function was executed ...
Probably it's ok to sote shard_ptr to State

Reply:
You mean `auto event = std::move(state);` this line?

I guess it was written to avoid leaving `state` in the `Job`, in `ThreadPoolImpl<Thread>::worker`, but it is reseted just after execution - https://github.com/ClickHouse/ClickHouse/blob/71aba5b4c663716752a1677d0971f576760648fd/src/Common/ThreadPool.cpp#L284

Although I guess w/o moving std::shared_ptr into the callback it's lifetime can be slightly greater, and it is better to move it into the callback.

Is that what you meant?